src/celeritas/user/DetectorSteps.cu

```diff
@@ -8,96 +8,123 @@
+    {
+        DS_COPY_IF_SELECTED(points[sp].time);
+        DS_COPY_IF_SELECTED(points[sp].pos);
+        DS_COPY_IF_SELECTED(points[sp].dir);
+        DS_COPY_IF_SELECTED(points[sp].energy);
     }
-};
 
+    DS_COPY_IF_SELECTED(event_id);
+    DS_COPY_IF_SELECTED(parent_id);
+    DS_COPY_IF_SELECTED(track_step_count);
+    DS_COPY_IF_SELECTED(step_length);
+    DS_COPY_IF_SELECTED(particle);
+    DS_COPY_IF_SELECTED(energy_deposition);
+#undef DS_COPY_IF_SELECTED
+}
+
+//---------------------------------------------------------------------------//
+// KERNEL INTERFACE
 //---------------------------------------------------------------------------//
-size_type count_num_valid(DVecDetector const& orig_ids)
+/*!
+ * Gather results from active tracks that are in a detector.
+ */
+void gather_step(DeviceRef<StepStateData> const& state, size_type size)
 {
-    return thrust::count_if(
-        thrust::device, orig_ids.begin(), orig_ids.end(), HasDetector{});
+    if (size == 0)
+    {
+        // No valid tracks
+        return;
+    }
+
+    CELER_LAUNCH_KERNEL(gather_step,
+                        celeritas::device().default_block_size(),
+                        size,
+                        celeritas::device().stream(state.stream_id).get(),
+                        state,
+                        size);
 }
 
 //---------------------------------------------------------------------------//
-// Simplify opaque ID pointers to reduce code bloat from identical
-// instantiations
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
 template<class T>
-struct PointerTransformer
-{
-    constexpr T* operator()(T* v) const { return v; }
-};
+using StateRef
+    = celeritas::StateCollection<T, Ownership::reference, MemSpace::device>;
 
-template<class V, class S>
-struct PointerTransformer<OpaqueId<V, S>>
+//---------------------------------------------------------------------------//
+struct HasDetector
 {
-    constexpr S* operator()(OpaqueId<V, S>* v) const
+    CELER_FORCEINLINE_FUNCTION bool operator()(DetectorId const& d)
     {
-        return reinterpret_cast<S*>(v);
+        return static_cast<bool>(d);
     }
 };
 
 //---------------------------------------------------------------------------//
 template<class T>
-void assign_field(std::vector<T>* dst,
-                  StateRef<T> const& src,
-                  DVecDetector const& orig_ids,
-                  size_type size,
-                  StreamId stream_id)
+void copy_field(std::vector<T>* dst, StateRef<T> const& src, size_type size)
 {
-    if (src.empty())
+    if (src.empty() || size == 0)
     {
         // This attribute is not in use
         dst->clear();
         return;
     }
-
-    PointerTransformer<T> transform_ptr;
-
-    // Partition based on detector validity
-    auto temp_span = src[AllItems<T>{}];
-    thrust::partition(
-        thrust::device.on(celeritas::device().stream(stream_id).get()),
```

Comment:
You said that using async memory copies didn't improve the transfers, right? I'm guessing though that host destination memory was allocated using the regular std::vector, but I remembered that CUDA requires [page-locked allocations](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gb65da58f444e7230d3322b6126bb4902) or "pinned" memory for async transfers to actually be async. For testing purposes we could see what the speedup is by replacing the underlying `std::vector` with a cuda host malloc everywhere; for a production solution we could maybe have a separate ownership enum: `asyncvalue` or something? Either way that would belong in a separate PR.

Reply:
or use [cudaHostRegister](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ge8d5c17670f16ac4fc8fcb4181cb490c) which "Page-locks the memory range specified by ptr and size", which might be way easier to do. (Maybe add a `ScopedPageLock<W, M>` which for host+value takes a `void*` constructor and calls `cudaHostRegister`, `cudaHostUnregister`?

Reply:
Oh awesome. Absolutely, I was trying to say that the async should be a separate commit (if at all), and clearly the async isn't low-hanging fruit (unless the blocking copies kill us at 32 threads, and assuming we care about scaling that high).